### ISSUE TITLE
Readme feedback

### ISSUE THREAD

**Issue by mlissner** [rachlllg/FLP_Semantic_Search#1]
I had a little time tonight and read through your readme. A few comments:

1. This is great. I appreciate the work you did organizing all this, putting together research, and understanding our needs.

2. I had a bunch of questions about re-ranking and how it works in practice, which I chatted with OpenAI about here (some of it's a bit silly; I can't help myself): 

    https://chatgpt.com/share/67b2c86c-341c-8013-ba34-5fa925a24346

3. If you want more examples of semantic search, you should check the notes in our CRM for others that have built similar systems that might be interested in chatting with us (a number of folks have expressed an interest in sharing their systems). Everybody that's interested, should be linked to this issue: 

    https://github.com/freelawproject/courtlistener/issues/5758

5. We know many of the folks that operate the prior art you've found, and we can consider intros, if you want.

6. You say:

    > Stats: 953 opinions selected (~1% of 9.7M opinions),

    I'm not one to judge if that's a good amount of records, but 1% of 10M is 100k, not 1k.

7. You say:

    > Even though gpt-4o-mini is not open source, given a good data is the cornerstone of a successful project, I figured a couple bucks is worth it.

    I think we chatted about this since you wrote this up, but just to be clear, it's fine to use closed source models. The big thing we need to avoid is sending user data to such models.

8. You say:

    > Consider adding other aspects of the filing in addition to the opinion for semantic search results

    That's interesting, and parentheticals could help too. But I have a question: If we do this, and we get back a result, can we highlight the relevant part of it? If it's just the text that matches, I assume we can do highlighting in that. If it can be one of several fields, can we still do it somehow? UI seems complicated, but I'm also curious how the back end would handle this.

9. You say:

    > also consider a combination of key work search with semantic reranking

    Yeah, that occurred to me too. The relevancy engine for keyword search uses a bunch of factors: 

     - BM25 to begin
     - Field boosting: If it matches the title, great, that's important.
     - *Extra* field boosting: If the query seems to be a case name, make the case name boost more powerful. 
     - Network boosting: Boost based on a network algorithm like pagerank (this is in the works).
     - Relevance decay: Older stuff is less important; demote it.
     - Jurisdiction boosting: Boost SCOTUS stuff, appellate stuff, etc. (this is planned)
     - Phrase boosting: If two keywords match a phrase, boost.
    
    And probably a few other things too, but those are the highlights.

That's it for now. Thanks for the detailed research and excellent report.

**Comment by rachlllg** [comment]
Thanks Mike!

> I had a bunch of questions about re-ranking and how it works in practice, which I chatted with OpenAI about here (some of it's a bit silly; I can't help myself)

Fun discussions, happy to discuss on the topic if you would like. Is this the Pro version of ChatGPT? Seems more advanced than the free version (with links & diagrams etc)

A few things:
1. In semantic search, we would most likely retrieve the top-K results (this is one of the parameters we specify in Elastic), brute-force search by retrieving all potentially relevant results is computationally impractical and the results after top-K are semantically too far away from the query to provide value.
2. We are using an approximate nearest neighbor (ANN) to retrieve the top-K results (this is provided as built-in from Elastic). The most basic ANN is clustering the embeddings and retrieve based on the centroids of the clusters, say we have 10M records with 1K clusters, this allows the algorithm to retrieve from 1K potential matches instead of 10M. Elastic uses a different algorithm (Hierarchical Navigable Small World graphs), but the goal is the same - efficient retrieval.
3. The retrieval model itself is (most often) a two-tower model, not just the reranker. Two-tower means you have two-encoders, one for the query, another for the context. [Here is a nice readme](https://sbert.net/examples/applications/cross-encoder/README.html) about it. You can see on the Bi-Encoder diagram, the two-towers are joined together to produce the final cosine-similarity. To train two-tower models, you need query-context pairs. 
4. Reranker is a model trained similar to that of the retrieval model, you still provide query-context pairs, the only difference is the objective function (also known as the loss function). While the retrieval model is trained to provide all top-K results, the reranker is trained to "sort" the results to put the ones closest to the query at the top. There are rerankers already trained that we can use out-of-the-box.

> you should check the notes in our CRM for others that have built similar systems that might be interested in chatting with us

This is great, thanks for sharing! I think most urgently, we need a golden dataset, this is query-context pairs of user-queries and best/worst-matched opinions for the given query. This dataset will enable us to evaluate our and others systems systematically and objectively. If gathering actual user queries and matched opinions may be time-consuming, perhaps we could consider having SME come up with queries and provide the best-matched opinions for the given queries. This may be a months-long venture, to polish out and create a nice dataset for finetuning and evaluation.
https://github.com/freelawproject/ai-research/issues/54

> I'm not one to judge if that's a good amount of records, but 1% of 10M is 100k, not 1k.

Thanks for catching this, it's a typo. I fixed it. I think the size here is not the most critical, but rather the distribution. In training a model, we want to ensure the dataset is balanced, such that the model learns both relevant and irrelevant examples and learns from all courts/jurisdictions equally. But in evaluating the model, we want to ensure the dataset is representative of the population, such that we know how the model would actually perform when deployed to the "wild".

We'll want to keep this in mind when we create the golden dataset.

> I think we chatted about this since you wrote this up, but just to be clear, it's fine to use closed source models. The big thing we need to avoid is sending user data to such models.

Yes, thanks for bringing this up. I emphasized in the `Model & Evaluation` section on why we chose an open source model - privacy and security. This note here is purely because I had emphasized this exercise is focused on open source, which GPT is clearly not. And GPT is used here to generate the synthetic dataset, not to be used for the retrieval.

> If we do this, and we get back a result, can we highlight the relevant part of it?

If we embed different aspects of the opinion (opinion, headmatter, parenthetical, etc) separately, we would be able to say "this retrieval is based on the embedding from this aspect of the opinion". I'm not sure if we'd be able to highlight a specific section of an opinion if the opinion is represented by one embedding. I'll have to look into this more, but as you said, complicated.

> Yeah, that occurred to me too. The relevancy engine for keyword search uses a bunch of factors:

Thank you for this detailed list, super helpful!

**Comment by mlissner** [comment]
> Is this the Pro version of ChatGPT? Seems more advanced than the free version (with links & diagrams etc)

Yeah, I think I used the reasoning model, but it's kind of funny, even as somebody that follows this stuff, I can't keep track of 1o vs. 4o, vs the other weird names they've come up with. So yeah, I think so? 

> I'm not sure if we'd be able to highlight a specific section 

Yeah, I'm not sure how this works. Maybe Elastic has something, but I noticed that the Casetext demo has some highlighting and it made me wonder how we'd do similar.

The rest sounds good!
